checker-util/src/main/java/org/checkerframework/checker/regex/util/RegexUtil.java

```diff
@@ -304,25 +308,66 @@ public static boolean isRegex(final char c) {
-     * given number of groups, otherwise throws an error. The purpose of this method is to suppress
-     * Regex Checker warnings. It should be very rarely needed.
+     * Returns the argument as a {@code @RegexNNGroups(groups = groups) String} if it is a regex
+     * with at least the given number of groups, otherwise throws an error. The purpose of this
+     * method is to suppress Regex Checker warnings. It should be very rarely needed.
      *
      * @param s string to check for being a regular expression
      * @param groups number of groups expected
      * @return its argument
-     * @throws Error if argument is not a regex
+     * @throws Error if argument is not a regex with at least the given number of groups
      */
     @SuppressWarnings("regex") // RegexUtil
     @SideEffectFree
     // The return type annotation is irrelevant; it is special-cased by
     // RegexAnnotatedTypeFactory.
-    public static @Regex String asRegex(String s, int groups) {
+    public static @RegexNNGroups String asRegex(String s, int groups) {
+        try {
+            Pattern p = Pattern.compile(s);
+            int actualGroups = getGroupCount(p);
+            if (actualGroups < groups) {
+                throw new Error(regexErrorMessage(s, groups, actualGroups));
+            }
+            return s;
+        } catch (PatternSyntaxException e) {
+            throw new Error(e);
+        }
+    }
+
+    /**
+     * Returns the argument as a {@code @RegexNNGroups(groups = groups, nonNullGroups =
+     * nonNullGroups} if it is a regex with at least the given number of groups and the groups in
+     * {@code nonNullGroups} are guaranteed to match provided that the regex matches a string,
+     * otherwise throws an error. The purpose of this method is to suppress Regex Checker warnings.
+     * It should be rarely needed.
+     *
+     * @param s string to check for being a regular expression
+     * @param groups number of groups expected
+     * @param nonNullGroups groups expected to be match some part of a target string when the regex
+     *     matches
+     * @return its argument
+     * @throws Error if argument is not a regex with the specified characteristics
+     */
+    @SuppressWarnings("regex")
+    @SideEffectFree
+    public static @RegexNNGroups String asRegex(String s, int groups, int... nonNullGroups) {
         try {
             Pattern p = Pattern.compile(s);
             int actualGroups = getGroupCount(p);
+            List<Integer> actualNonNullGroups = getNonNullGroups(p.pattern(), actualGroups);
+            boolean containsAll = true;
+            int failingGroup = -1;
+            for (int e : nonNullGroups) {
+                if (!actualNonNullGroups.contains(e)) {
+                    containsAll = false;
+                    failingGroup = e;
+                    break;
+                }
+            }
             if (actualGroups < groups) {
```

Comment:
I think the method should do this simpler, cheaper test before the test about the non-null groups.